oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java

```diff
@@ -0,0 +1,308 @@
+            long readPos = 0;
+            int outFileIndex = 1;
+            File currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(algorithm));
+            BufferedWriter writer = createWriter(currentFile, algorithm);
+            try {
+                splitFlatFiles.add(currentFile);
+                int lineCount = 0;
+                String line;
+                Stack<String> nodeTypeNameStack = new Stack<>();
+                while ((line = reader.readLine()) != null) {
+                    updateNodeTypeStack(nodeTypeNameStack, line);
+                    boolean shouldSplit = (readPos > splitThreshold);
+                    if (shouldSplit && canSplit(indexNodeTypesName, nodeTypeNameStack)) {
+                        writer.close();
+                        LOG.info("Created split flat file {} with size {}", currentFile.getAbsolutePath(), FileUtils.byteCountToDisplaySize(currentFile.length()));
+                        readPos = 0;
+                        outFileIndex++;
+                        currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(algorithm));
+                        writer = createWriter(currentFile, algorithm);
+                        splitFlatFiles.add(currentFile);
+                        LOG.info("Split position found at line {}, creating new split file {}", lineCount, currentFile.getAbsolutePath());
+                    }
+                    writer.append(line);
+                    writer.newLine();
+                    readPos += line.length() + 1;
+                    lineCount++;
+                }
+                LOG.info("Created split flat file {}, size {}, line count {}",
+                        currentFile.getAbsolutePath(),
+                        FileUtils.byteCountToDisplaySize(currentFile.length()),
+                        lineCount);
+            } finally {
+                writer.close();
+            }
+        }
+
+        if (deleteOriginal) {
+            LOG.info("Removing original flat file {} after splitting into {} files", flatFile.getAbsolutePath(), splitFlatFiles);
+            flatFile.delete();
+        }
+
+        return splitFlatFiles;
+    }
+
+    /**
+     * Update the stack of node types. The first entry is the node type of the root
+     * node, then the node type of the child,... up to the node type of the current
+     * node.
+     * 
+     * @param parentNodeTypeNames the stack of node types processed so far
+     * @param line the path of the current node
+     */
+    private void updateNodeTypeStack(Stack<String> parentNodeTypeNames, String line) {
+        NodeStateHolder ns = new SimpleNodeStateHolder(line);
+        List<String> pathElements = ns.getPathElements();
+        int currentLineDepth = pathElements.size();
+        int parentTypesDepth = parentNodeTypeNames.size();
+        if (currentLineDepth > parentTypesDepth) {
+            // TODO we also need to consider mixins
```

Comment:
(I added this TODO on purpose: we can't merge the PR currently, because this still needs to be implemented.)

The code only deals with primary types (e.g. acme:Asset), but doesn't deal with mixing types.
We need to work on this, before we can merge the patch.
Using a stack alone isn't sufficient: we need a stack of sets :-/